Assignment3/Q1.py

```diff
@@ -0,0 +1,128 @@
+"""The approach taken in this solution is based on a DFS-like search. The worst case running time
+   is exponential but because of the optimization used with isPrefix(string) method the average
+   running time should be a lot better.
+   The solution given below uses create_corresponding_graph(rows, columns) method to create a graph corresponding
+   to the grid - 2 dimensional array given. It uses a adjacency list representation of a graph
+   (Python dictionary) with an edge between every corresponding two adjacent entries in a given grid.
+   Then using both the given grid and a corresponding graph the DFS-like search is performed on a graph
+   (using words_search_recursive method) but it searches ALL possible paths from each vertex (corresponding
+   to a letter) similar to creating a big search tree with all possible words. Using isPrefix(string) method
+   it 'cuts off' early all the branches of the search tree which is being created that cannot form words."""
+
+from Dictionary import Dictionary
+
+
+def create_corresponding_graph(rows, columns):
+    """Method takes two arguments and returns a adjacency list representation of the corresponding grid with given
+       number of rows and columns.
+
+        Args:
+            rows: integer, number of rows in the grid
+            columns: integer, number of columns in the grid
+
+        Returns:
+            a dictionary, adjacency list representation of the graph corresponding to the character grid given in
+            the problem. A vertex in a graph is an integer tuple which indicates the coordinates of an entry in
+            the grid.
+    """
+    # Using a dictionary data structure for the adjacency list representation of the graph.
+    graph = {}
+    for i in range(rows):
+        for j in range(columns):
+            # Add an entry (tuple of coordinates) for each cell in a grid.
+            graph[(i, j)] = []
+
+            # Create arrays of possible adjacent entries.
+            adjacent_horizontal = [i - 1, i, i + 1]
+            adjacent_vertical = [j - 1, j, j + 1]
+
+            # Populate each entry in a graph adjacency list representation.
+            for h in adjacent_horizontal:
+                for v in adjacent_vertical:
+                    # There are eight adjacent cells to each cell in a grid
+                    # Apart from edge cases, here we take care of edge cases.
+                    if 0 <= h < rows and 0 <= v < columns and not (h == i and v == j):
+                        graph[(i, j)] = graph[(i, j)] + [(h, v)]
+
+    # Return an adjacency list representation of a graph corresponding to a grid.
+    return graph
+
+
+def word_search(rows, columns, grid, dictionary):
+    """A method performs a word search on a character grid and returns a set of words (no duplicates) that can be formed
+       from grid's characters and are in a dictionary. Word search is case insensitive.
+
+        Args:
+            rows: an integer, number of rows in a grid
```

Comment:
Ohh, I thought we should pass them as an argument as it was mentioned in an assignment question, changed now!